File: mnist/fnn/R/fnn.R
cross_entropy = function(p, q){
    ## NOTE (Michael): This is to avoid the cross-entropy to be undefined when q
    ##                 is 1 or 0. However, there should be a better way of
    ##                 dealing with this.
    k = 1e-10
    sq = ifelse(q == 0, k, ifelse(q == 1, 1 - k, q))
    s = p * log(sq) + (1 - p) * log(1 - sq)
    -sum(s) * (1/NROW(q))
}

sigmoid = function(x){
    1/(1 + exp(-x))
}

translation = function(x, w, b){
    t(t(x %*% w) + c(b))
}

cross_entropy_delta = function(p, q){
    k = 1e-15
    sq = ifelse(q == 0, k, ifelse(q == 1, 1 - k, q))
    (p - sq)/(sq * (sq - 1))/(length(q))
}

sigmoid_delta = function(x){
    delta = exp(-x)/((1 + exp(-x))^2)
    ## NOTE (Michael): This is to avoid numerical error
    delta[is.nan(delta)] = 0
    delta
}

translation_delta_weight= function(x){
    t(x)
}

translation_delta_bias = function(x){
    matrix(1, nr = 1, nc = nrow(x))
}

translation_delta_x = function(w){
    t(w)
}


initialise_weights = function(size){
    layers = length(size) - 1
    weights = vector(mode = "list", layers)
    for(layer in 1:layers){
        ## weights[[layer]] =
        ##     apply(X = matrix(rnorm(size[layer] * size[layer + 1]),
        ##                      nr = size[layer],
        ##                      nc = size[layer + 1]),
        ##           MARGIN = 2,
        ##           FUN = function(x) x/sum(x))
        weights[[layer]] =
            matrix(rnorm(size[layer] * size[layer + 1]),
                   nr = size[layer],
                   nc = size[layer + 1])
    }
    weights
}

initialise_bias = function(size){
    layers = length(size) - 1
    bias = vector(mode = "list", layers)
    for(layer in 1:layers){
        bias[[layer]] = matrix(rnorm(size[layer + 1]),
                               nc = 1, nr = size[layer + 1])
    }
    bias
}



fp = function(data, weights, bias, activationFUN){

    n.layers = length(weights) + 1

    translation_layer = vector("list", n.layers)
    translation_layer[[1]] = data
    activation_layer = translation_layer


    for(layer in 1:(n.layers - 1)){
        translation_layer[[layer + 1]] =
            translation(translation_layer[[layer]], weights[[layer]], bias[[layer]])
        activation_layer[[layer + 1]] =
            activationFUN(translation_layer[[layer + 1]])
    }
    list(translation_layer = translation_layer,
         activation_layer = activation_layer)
}

bp = function(label, weights, bias, fp, costDerivFUN, activationDerivFUN, gamma){
    n.layers = length(weights) + 1
    w_tmp = weights
    b_tmp = bias
    dt_dw = vector("list", n.layers - 1)
    dt_db = vector("list", n.layers - 1)
    dt_da = vector("list", n.layers - 2)
    da_dt = vector("list", n.layers - 1)

    dc_da =
        costDerivFUN(label, fp$activation_layer[[n.layers]])
    da_dt[[n.layers - 1]] =
        activationDerivFUN(fp$translation_layer[[n.layers]])
    dc_dt = dc_da * da_dt[[n.layers - 1]]
    dt_dw[[n.layers - 1]] =
        t(translation_delta_weight(fp$activation_layer[[n.layers - 1]]))
    dt_db[[n.layers - 1]] =
        t(translation_delta_bias(fp$activation_layer[[n.layers - 1]]))
    w_tmp[[n.layers - 1]] =
        w_tmp[[n.layers - 1]] - gamma * t(t(dc_dt) %*% dt_dw[[n.layers - 1]])
    b_tmp[[n.layers - 1]] =
        b_tmp[[n.layers - 1]] - gamma * t(dc_dt) %*% dt_db[[n.layers - 1]]

    for(layer in (n.layers - 2):1){
        ## Calculate individual derivatives
        da_dt[[layer]] =
            activationDerivFUN(fp$translation_layer[[layer + 1]])
        dt_dw[[layer]] =
            t(translation_delta_weight(fp$activation_layer[[layer]]))
        dt_db[[layer]] =
            t(translation_delta_bias(fp$activation_layer[[layer]]))
        dt_da[[layer]] =
            translation_delta_x(weights[[layer + 1]])

        ## Update derivative of cost to each layer
        dc_dt = (dc_dt %*% dt_da[[layer]]) * da_dt[[layer]]

        ## Update the weights
        w_tmp[[layer]] =
            w_tmp[[layer]] - gamma * t(t(dc_dt) %*% dt_dw[[layer]])
        b_tmp[[layer]] =
            b_tmp[[layer]] - gamma * t(dc_dt) %*% dt_db[[layer]]
    }
    list(weights = w_tmp, bias = b_tmp)
}

sgd = function(data,
               label,
               size,
               costFUN,
               activationFUN,
               costDerivFUN,
               activationDerivFUN,
               maxIter,
               gamma,
               batchSize = 50){

    ## Initialisation
    n_data= nrow(data)
    n.layers = length(size)
    weights = initialise_weights(size)
    bias = initialise_bias(size)
    i = 1

    ## Start stochastic gradient descent
    while(i <= maxIter){
        shuffleIndex = sample(n_data, n_data)
        shuffledData = data[shuffleIndex, ]
        shuffledLabel = label[shuffleIndex, ]
        n_sample = floor(n_data * 1/batchSize)
        for(batch in 1:batchSize){
            index_sample = 1:n_sample + ((batch - 1) * n_sample)
            train_data = shuffledData[index_sample, ]
            train_data_label = shuffledLabel[index_sample, ]

            ## Forward propagation
            forward = fp(data = train_data,
                         weights = weights,
                         bias = bias,
                         activationFUN = activationFUN)

            ## TODO (Michael): Need a way to identify the convergence of the
            ##                 stochastic gradient descent.

            ## Back propagation
            back = bp(label = train_data_label,
                      weights = weights,
                      bias = bias,
                      fp = forward,
                      costDerivFUN = costDerivFUN,
                      activationDerivFUN = activationDerivFUN,
                      gamma)
            weights = back$weights
            bias = back$bias

        }

        finalForward =
            fp(data = shuffledData,
               weights = weights,
               bias = bias,
               activationFUN = activationFUN)
        cost =
            costFUN(shuffledLabel, finalForward$activation_layer[[n.layers]])
        message("Cost: ", cost)

        ## Calculate the number correctly classified
        pred = apply(finalForward$activation_layer[[n.layers]], 1, which.max) - 1
        actual_label = apply(shuffledLabel, 1, which.max) - 1
        message("Percentage classified correctly: ",
                round(sum(actual_label == pred)/n_data * 100, 4), "%")

        ## Increment i
        i = i + 1

        ## ## Shrink gamma
        ## gamma = gamma * 0.98
        ## print(gamma)

    }
    list(weights = weights, bias = bias)
}

rsgd = function(data,
                label,
                size,
                costFUN,
                activationFUN,
                costDerivFUN,
                activationDerivFUN,
                maxIter,
                gamma,
                batchSize = 50,
                samplingPct = 0.3){

    ## Initialisation
    n_data= nrow(data)
    n.layers = length(size)
    weights = initialise_weights(size)
    bias = initialise_bias(size)

    i = 1
    n_sample = floor(n_data * samplingPct)
    while(i <= maxIter){

        index_sample = sample(n_data, n_sample)
        train_data = data[index_sample, ]
        train_data_label = label[index_sample, ]

        ## Forward propagation
        forward = fp(data = train_data,
                     weights = weights,
                     bias = bias,
                     activationFUN = activationFUN)
        cost =
            costFUN(train_data_label, forward$activation_layer[[n.layers]])

        message("Cost: ", cost)

        ## TODO (Michael): Need a way to identify the convergence of the
        ##                 stochastic gradient descent.

        ## Back propagation
        back = bp(label = train_data_label,
                  weights = weights,
                  bias = bias,
                  fp = forward,
                  costDerivFUN = costDerivFUN,
                  activationDerivFUN = activationDerivFUN,
                  gamma)
        bias = back$bias
        weights = back$weights

        ## Calculate the number correctly classified
        pred = apply(forward$activation_layer[[n.layers]], 1, which.max) - 1
        actual_label = apply(train_data_label, 1, which.max) - 1
        message("Percentage classified correctly: ",
                round(sum(actual_label == pred)/n_sample * 100, 4), "%")

        ## Increment i
        i = i + 1

    }
    list(weights = weights, bias = bias)
}



fnn = function(data,
               label,
               size,
               costFUN,
               activationFUN,
               costDerivFUN,
               activationDerivFUN,
               maxIter,
               tol,
               stochastic = TRUE,
               gamma = 1e-3,
               batchSize = 20,
               samplingPct){

    ## Check
    if(ncol(data) != size[1])
        stop("Incorrect input size")
    if(ncol(label) != size[length(size)])
        stop("Incorrect output size")

    ## Estimate the model
    params =
        sgd(data = data,
            label = label,
            size = size,
            costFUN = costFUN,
            activationFUN = activationFUN,
            activationDerivFUN = activationDerivFUN,
            costDerivFUN = costDerivFUN,
            maxIter = maxIter,
            gamma = gamma,
            batchSize = batchSize)

    ## params =
    ##     rsgd(data = data,
    ##          label = label,
    ##          size = size,
    ##          costFUN = costFUN,
    ##          activationFUN = activationFUN,
    ##          activationDerivFUN = activationDerivFUN,
    ##          costDerivFUN = costDerivFUN,
    ##          maxIter = maxIter,
    ##          gamma = gamma,
    ##          batchSize = batchSize,
    ##          samplingPct = samplingPct)

    ## Return model
    model = list(model_data = data,
                 label = label,
                 size = size,
                 costFUN = costFUN,
                 activationFUN = activationFUN,
                 weights = params$weights,
                 bias = params$bias)
    class(model) = "fnn"
    model
}

predict = function(data, model){
    n.layers = length(model$size)
    forward =
        with(model,
             fp(data, model$weights, model$bias, activationFUN))
    forward$activation_layer[[n.layers]]
}

File: mnist/fnn/python/fnn.py
import numpy as np
import random

def sigmoid(z):
    """The sigmoid function."""
    return 1.0/(1.0+np.exp(-z))

def sigmoid_delta(z):
    """Derivative of the sigmoid function."""
    return sigmoid(z)*(1-sigmoid(z))

def cross_entropy(p, q):
    """The Cross-entropy function."""
    ## This is to avoid numerical error as the entropy is undefined when q is
    ## either 0 or 1.
    k = 1e-10
    q[q == 0] = k
    q[q == 1.0] = 1.0 - k
    cross_entropy = p * np.log(q) + (1.0 - p) * np.log(1.0 - q)
    return -np.sum(cross_entropy) * 1.0/q.size


def cross_entropy_delta(p, q):
    """The derivative of the cross entropy function"""
    k = 1e-10
    q[q == 0] = k
    q[q == 1.0] = 1.0 - k
    return (p - q)/(q * (q - 1.0))/q.size


def trans(x, w, b):
    return np.dot(x, w) + b

def trans_delta_weights(x):
    return x.T

def trans_delta_biases(x):
    return np.ones(x.shape[1])

def trans_delta_activation(w):
    return w.T





class Fnn(object):

    def __init__(self, size):
        self.size = size
        self.num_layer = len(size)
        self.biases = [np.random.randn(1, nodes) for nodes in self.size[1:]]
        self.weights = [np.random.randn(input, output)
                        for input, output in zip(size[:-1], size[1:])]
        self.activation = [None] * (self.num_layer)
        self.trans = [None] * (self.num_layer - 1)

    def feedforward(self, x, y):
        """Performs the feedforward, the transformation and the activation layers
        are updated.

        """
        ## Initialise, although we can assign x to activation and then change
        ## the index
        self.activation[0] = x

        for layer in range(0, self.num_layer - 1):
            ## Perform transformation
            self.trans[layer] = np.dot(self.activation[layer],
                                       self.weights[layer]) + \
                                       self.biases[layer]
            ## Perform activation
            self.activation[layer + 1] = sigmoid(self.trans[layer])

        cost = cross_entropy(y, self.activation[-1])
        if(np.isnan(cost)):
            raise Exception("Cost is nan")
        print(cost)

    def backpropagation(self, x, y, gamma):
        """Take the activation and trasformation layer, and updates the biases and
        weights

        """
        ## NOTE (Michael): There still seems to be something wrong with the back
        ##                 propagation. Initialise the nabla (gradient)
        nabla_weights = [np.zeros(w.shape) for w in self.weights]
        nabla_biases = [np.zeros(b.shape) for b in self.biases]

        ## Derivative of cost w.r.t to transformation
        delta = cross_entropy_delta(y, self.activation[-1]) * \
                sigmoid_delta(self.trans[-1])
        ## Derivative of transformation to weights
        nabla_weights[-1] = np.dot(trans_delta_weights(self.activation[-2]),
                                   delta)
        ## Derivative of trans to bias
        nabla_biases[-1] = delta
        for layer in xrange(2, self.num_layer):
            delta = np.dot(delta,
                           trans_delta_activation(self.weights[-layer + 1]))
            delta = delta * sigmoid_delta(self.trans[-layer])
            nabla_weights[-layer] = np.dot(trans_delta_weights(self.activation[-layer - 1]),
                                           delta)
            nabla_biases[-layer] = delta

        self.weights = [w - gamma * nw
                        for w, nw in zip(self.weights, nabla_weights)]
        self.biases = [b - gamma * nb
                       for b, nb in zip(self.biases, nabla_biases)]

    def train_mini_batch(self, mini_batch, gamma):
        """Input is the shuffled data, and calls the feedforward and backpropagation
        method to update the biase, weights, activation and transformation.

        """
        for sx, sy in mini_batch:
            ## Do feedforward
            self.feedforward(sx.T, sy.T)
            ## Do back propagation
            self.backpropagation(sx.T, sy.T, gamma)


    def train(self, training_data, epochs, mini_batch_size, gamma, test_data = None):
        """Train the model"""
        for epoch in xrange(epochs):
            ## Shuffle the data
            random.shuffle(training_data)
            n = len(training_data)
            ## Create the mini batches
            mini_batches = [
                training_data[k:k + mini_batch_size]
                for k in xrange(0, n, mini_batch_size)]
            ## train with each mini batch using back propagation
            for mini_batch in mini_batches:
                self.train_mini_batch(mini_batch, gamma)



import mnist_loader
training_data, validation_data, test_data = mnist_loader.load_data_wrapper('/home/mk/Github/deep_learning/mnist/data/mnist.pkl.gz')

size = [784, 30, 10]
## Model 1
net = Fnn(size)
net.train(training_data, 30, 10, 0.1, test_data=test_data)
